src/gluonts/model/forecast_generator.py

```diff
@@ -0,0 +1,222 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import logging
+from typing import Any, Callable, Iterator, List, Optional
+
+# Third-party imports
+import mxnet as mx
+import numpy as np
+
+# First-party imports
+from gluonts.distribution import Distribution, DistributionOutput
+from gluonts.core.component import validated
+from gluonts.dataset.common import DataEntry
+from gluonts.dataset.field_names import FieldName
+from gluonts.dataset.loader import InferenceDataLoader
+from gluonts.model.forecast import (
+    Forecast,
+    SampleForecast,
+    QuantileForecast,
+    DistributionForecast,
+)
+
+
+OutputTransform = Callable[[DataEntry, np.ndarray], np.ndarray]
+BlockType = mx.gluon.Block
+
+
+LOG_CACHE = set([])
+
+
+def log_once(msg):
+    global LOG_CACHE
+    if msg not in LOG_CACHE:
+        logging.info(msg)
+        LOG_CACHE.add(msg)
+
+
+def _extract_instances(x: Any) -> Any:
+    """
+    Helper function to extract individual instances from batched
+    mxnet results.
+
+    For a tensor `a`
+      _extract_instances(a) -> [a[0], a[1], ...]
+
+    For (nested) tuples of tensors `(a, (b, c))`
+      _extract_instances((a, (b, c)) -> [(a[0], (b[0], c[0])), (a[1], (b[1], c[1])), ...]
+    """
+    if isinstance(x, (np.ndarray, mx.nd.NDArray)):
+        for i in range(x.shape[0]):
+            # yield x[i: i + 1]
+            yield x[i]
+    elif isinstance(x, tuple):
+        for m in zip(*[_extract_instances(y) for y in x]):
+            yield tuple([r for r in m])
+    elif x is None:
+        while True:
+            yield None
```

Comment:
What is the purpose of this? Shouldn't the `x is None` also throw an exception?

Reply:
This handles the case where one of the return tensors is `None`, which mxnet supports. Then the function extracts things correctly, with a `None` value in the corresponding entry of each instance. In the specific case of the `DistributionForecast` the `scale` can be none.

Reply:
I see, so this is primarily for the case of tuples having `None` among the elements -- let's make sure `_extract_instances` is never invoked with `None` at the top level :-)